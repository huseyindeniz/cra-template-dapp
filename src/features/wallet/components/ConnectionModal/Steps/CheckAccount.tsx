import { Box, VStack, Button, Text, Progress } from '@chakra-ui/react';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { AlertMessage } from '../../../../ui/components/AlertMessage/AlertMessage';
import { WalletAccountStateType } from '../../../types';

export interface CheckAccountProps {
  stepState: WalletAccountStateType;
  errorMessage: string | null;
  onUnlock: () => void;
}

export const CheckAccount: React.FC<CheckAccountProps> = ({
  stepState,
  errorMessage,
  onUnlock,
}) => {
  const { t } = useTranslation('FeatureWallet');

  const AccountDetectionFailed = () => {
    return (
      <Box w="full">
        <AlertMessage status="warning" title={t('Unexpected Error')}>
          <Text fontSize="xs">
            {t('An error has occured during the wallet status check.')}
            <br /> {t('Please try again later.')}
            <br />
            {t('The error code was')}: {errorMessage}
          </Text>
        </AlertMessage>
      </Box>
    );
  };

  const Locked = () => {
    return (
      <Box w="full">
        <AlertMessage status="warning" title={t('Wallet Is Locked')}>
          <Text fontSize="xs">
            {t('Please unlock your wallet if you want to continue.')}
          </Text>
        </AlertMessage>
      </Box>
    );
  };

  const UnlockRequested = () => {
    return (
      <Box w="full">
        <Progress size="xs" isIndeterminate colorScheme="yellow" />
        <Text fontSize="xs">
          {t('Waiting for the unlock wallet request to be accepted.')}
          <br />
          {t('Please check your Metamask wallet.')}
        </Text>
      </Box>
    );
  };

  const UnlockRejected = () => {
    return (
      <Box w="full">
        <AlertMessage status="warning" title={t('Unlock Rejected')}>
          <Text fontSize="xs">
            {t('You rejected the unlock wallet request.')}
            <br />
            {t('Please try again if you want to continue.')}
          </Text>
        </AlertMessage>
      </Box>
    );
  };

  const UnlockFailed = () => {
    return (
      <Box w="full">
        <AlertMessage status="warning" title={t('Unexpected Error')}>
          <Text fontSize="xs">
            {t('An error has occured during the unlock wallet check.')}
            <br /> {t('Please try again later.')}
            <br />
            {t('The error code was')}: {errorMessage}
          </Text>
        </AlertMessage>
      </Box>
    );
  };

  const Content = () => {
    switch (stepState) {
      case WalletAccountStateType.ACCOUNT_DETECTION_FAILED:
        return <AccountDetectionFailed />;
      case WalletAccountStateType.LOCKED:
        return <Locked />;
      case WalletAccountStateType.UNLOCK_REQUESTED:
        return <UnlockRequested />;
      case WalletAccountStateType.UNLOCK_REJECTED:
        return <UnlockRejected />;
      case WalletAccountStateType.UNLOCK_FAILED:
        return <UnlockFailed />;
      default:
        return null;
    }
  };

  const UnlockButton = () => {
    switch (stepState) {
      case WalletAccountStateType.LOCKED:
      case WalletAccountStateType.UNLOCK_REJECTED:
        return (
          <Box>
            <Button
              variant="solid"
              colorScheme="yellow"
              onClick={() => onUnlock()}
            >
              {t('Unlock Wallet')}
            </Button>
          </Box>
        );
      default:
        return null;
    }
  };

  return (
    <VStack>
      <Content />
      <UnlockButton />
    </VStack>
  );
};
