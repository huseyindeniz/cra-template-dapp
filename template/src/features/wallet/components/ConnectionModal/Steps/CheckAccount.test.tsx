import { composeStories } from '@storybook/testing-react';
import { render } from '@testing-library/react';

import * as stories from './CheckAccount.stories';

describe('Feature: Wallet', () => {
  describe('Component: ConnectionModal/Steps/CheckAccount', () => {
    const {
      CheckAccountIdle,
      AccountRequested,
      AccountDetectionFailed,
      Locked,
      UnlockRequested,
      UnlockRejected,
      UnlockFailed,
      AccountLoaded,
    } = composeStories(stories);
    const onClickSpy = jest.fn();

    describe('Scenario: CheckAccountIdle, AccountRequested, AccountLoaded', () => {
      it('should be empty', () => {
        // Act
        const result1 = render(<CheckAccountIdle />);
        const result2 = render(<AccountRequested />);
        const result3 = render(<AccountLoaded />);
        // Assert
        expect(result1.container.childElementCount).toEqual(1);
        expect(result2.container.childElementCount).toEqual(1);
        expect(result3.container.childElementCount).toEqual(1);
      });
    });

    describe('Scenario: AccountDetectionFailed', () => {
      it('should be visible and show error', () => {
        // Arrange
        const expectedTitle = 'Unexpected Error';
        const expectedErrorMessage =
          AccountDetectionFailed.args?.errorMessage || '==not_exist==';
        // Act
        const { asFragment, getByText } = render(<AccountDetectionFailed />);
        // Assert
        const actualTitle = getByText(expectedTitle);
        const actualErrorMessage = getByText(
          new RegExp(expectedErrorMessage, 'i')
        );

        expect(actualErrorMessage).toHaveTextContent(expectedErrorMessage);
        expect(actualTitle).toBeVisible();
        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('Scenario: Locked', () => {
      it('should be visible and show error', () => {
        // Arrange
        const expectedTitle = 'Wallet Is Locked';
        // Act
        const { asFragment, getByText } = render(
          <Locked onUnlock={onClickSpy} />
        );
        getByText('Unlock Wallet').click();
        // Assert
        const actualTitle = getByText(expectedTitle);
        const actualButton = getByText('Unlock Wallet');

        expect(actualTitle).toBeVisible();
        expect(actualButton).toBeVisible();
        expect(onClickSpy).toHaveBeenCalled();
        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('Scenario: UnlockRequested', () => {
      it('should be visible and show error', () => {
        // Arrange
        const expectedMessage =
          'Waiting for the unlock wallet request to be accepted.';
        // Act
        const { asFragment, getByText } = render(<UnlockRequested />);
        // Assert
        const actualMessage = getByText(new RegExp(expectedMessage, 'i'));

        expect(actualMessage).toHaveTextContent(expectedMessage);
        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('Scenario: UnlockRejected', () => {
      it('should be visible and show error', () => {
        // Arrange
        const expectedTitle = 'Unlock Rejected';
        const expectedErrorMessage = 'You rejected the unlock wallet request.';
        // Act
        const { asFragment, getByText } = render(
          <UnlockRejected onUnlock={onClickSpy} />
        );
        getByText('Unlock Wallet').click();
        // Assert
        const actualTitle = getByText(expectedTitle);
        const actualErrorMessage = getByText(
          new RegExp(expectedErrorMessage, 'i')
        );

        expect(actualErrorMessage).toHaveTextContent(expectedErrorMessage);
        expect(actualTitle).toBeVisible();
        expect(onClickSpy).toHaveBeenCalled();
        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('Scenario: UnlockFailed', () => {
      it('should be visible and show error', () => {
        // Arrange
        const expectedTitle = 'Unexpected Error';
        const expectedErrorMessage =
          UnlockFailed.args?.errorMessage || '==not_exist==';
        // Act
        const { asFragment, getByText } = render(<UnlockFailed />);
        // Assert
        const actualTitle = getByText(expectedTitle);
        const actualErrorMessage = getByText(
          new RegExp(expectedErrorMessage, 'i')
        );

        expect(actualErrorMessage).toHaveTextContent(expectedErrorMessage);
        expect(actualTitle).toBeVisible();
        expect(asFragment).toMatchSnapshot();
      });
    });
  });
});
