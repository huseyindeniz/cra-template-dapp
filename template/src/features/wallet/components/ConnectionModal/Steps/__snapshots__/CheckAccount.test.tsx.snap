// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: Wallet Component: ConnectionModal/Steps/CheckAccount Scenario: AccountDetectionFailed should be visible and show error 1`] = `[Function]`;

exports[`Feature: Wallet Component: ConnectionModal/Steps/CheckAccount Scenario: Locked should be visible and show error 1`] = `[Function]`;

exports[`Feature: Wallet Component: ConnectionModal/Steps/CheckAccount Scenario: UnlockFailed should be visible and show error 1`] = `[Function]`;

exports[`Feature: Wallet Component: ConnectionModal/Steps/CheckAccount Scenario: UnlockRejected should be visible and show error 1`] = `[Function]`;

exports[`Feature: Wallet Component: ConnectionModal/Steps/CheckAccount Scenario: UnlockRequested should be visible and show error 1`] = `[Function]`;
